openmc/deplete/batchwise.py

```diff
@@ -0,0 +1,1157 @@
+        self._bracketed_method = value
+
+    @property
+    def tol(self):
+        return self._tol
+
+    @tol.setter
+    def tol(self, value):
+        check_type("tol", value, Real)
+        self._tol = value
+
+    @property
+    def target(self):
+        return self._target
+
+    @target.setter
+    def target(self, value):
+        check_type("target", value, Real)
+        self._target = value
+
+    @classmethod
+    def from_params(cls, obj, attr, operator, model, **kwargs):
+        return cls(obj, attr, operator, model, **kwargs)
+
+    @abstractmethod
+    def _model_builder(self, param):
+        """
+        Builds the parametric model to be passed to the
+        :meth:`openmc.search.search_for_keff` method.
+        Callable function which builds a model according to a passed
+        parameter. This function must return an openmc.model.Model object.
+        Parameters
+        ----------
+        param : parameter
+            model function variable
+        Returns
+        -------
+        _model :  openmc.model.Model
+            OpenMC parametric model
+        """
+
+    def _search_for_keff(self, val):
+        """
+        Perform the criticality search for a given parametric model.
+        If the solution lies off the initial bracket, this method iteratively
+        adapt it until :meth:`openmc.search.search_for_keff` return a valid
+        solution.
+        The adapting bracket algorithm takes the ratio between the bracket and
+        the returned keffs values to scale the bracket until a valid solution
+        is found.
+        A bracket limit pose the upper and lower boundaries to the adapting
+        bracket. If one limit is hit, the algorithm will stop and the closest
+        limit value will be used.
+        Parameters
+        ----------
+        val : float
+            Previous result value
+        Returns
+        -------
+        root : float
+            Estimated value of the variable parameter where keff is the
+            targeted value
+        """
+        #make sure we don't modify original bracket and tol values
+        bracket = deepcopy(self.bracket)
+
+        #search_for_keff tolerance should vary according to the first guess value
+        if abs(val) > 1.0:
+            tol = self.tol / abs(val)
+        else:
+            tol = self.tol
+
+        # Run until a search_for_keff root is found or out of limits
+        root = None
+
+        while root == None:
+            search = search_for_keff(self._model_builder,
+                            bracket = np.array(bracket) + val,
+                            tol = tol,
+                            bracketed_method = self.bracketed_method,
+                            target = self.target,
+                            print_iterations = self.print_iterations,
+                            run_args = {'output': self.search_for_keff_output},
+                            run_in_memory = True)
```

Comment:
I get nervous when I see a `while` loop that doesn't have a clear exit strategy. Consider the case where I want to run a critical boron search through depletion, but later in the simulation I've burned so much of my fuel that even with no boron I am subcritical. This loop would run forever.

Can we have the user provide an iteration limit, even defaulting to a high number, so we have a guaranteed exit condition?